sql-plugin/src/main/scala/org/apache/spark/sql/rapids/stringFunctions.scala

```diff
@@ -767,6 +771,595 @@ class GpuRLikeMeta(
       GpuRLike(lhs, rhs)
 }
 
+/**
+ * Regular expression parser based on a Pratt Parser design.
+ *
+ * The goal of this parser is to build a minimal AST that allows us
+ * to validate that we can support the expression on the GPU. The goal
+ * is not to parse with the level of detail that would be required if
+ * we were building an evaluation engine. For example, operator precedence is
+ * largely ignored but could be added if we need it later.
+ *
+ * The Java and cuDF regular expression documentation has been used as a reference:
+ *
+ * Java regex: https://docs.oracle.com/javase/7/docs/api/java/util/regex/Pattern.html
+ * cuDF regex: https://docs.rapids.ai/api/libcudf/stable/md_regex.html
+ *
+ * The following blog posts provide some background on Pratt Parsers and parsing regex.
+ *
+ * - https://journal.stuffwithstuff.com/2011/03/19/pratt-parsers-expression-parsing-made-easy/
+ * - https://matt.might.net/articles/parsing-regex-with-recursive-descent/
+ */
+class RegexParser(pattern: String) {
+
+  /** index of current position within the string being parsed */
+  private var i = 0
+
+  def parse(): RegexAST = {
+    val ast = parseInternal()
+    if (!eof()) {
+      throw new RegexUnsupportedException("failed to parse full regex")
+    }
+    ast
+  }
+
+  private def parseInternal(): RegexAST = {
+    val term = parseTerm(() => peek().contains('|'))
+    if (!eof() && peek().contains('|')) {
+      consumeExpected('|')
+      RegexChoice(term, parseInternal())
+    } else {
+      term
+    }
+  }
+
+  private def parseTerm(until: () => Boolean): RegexAST = {
+    val sequence = RegexSequence(new ListBuffer())
+    while (!eof() && !until()) {
+      parseFactor() match {
+        case RegexSequence(parts) =>
+          sequence.parts ++= parts
+        case other =>
+          sequence.parts += other
+      }
+    }
+    sequence
+  }
+
+  private def isValidQuantifierAhead(): Boolean = {
+    if (peek().contains('{')) {
+      val bookmark = i
+      consumeExpected('{')
+      val q = parseQuantifierOrLiteralBrace()
+      i = bookmark
+      q match {
+        case _: QuantifierFixedLength | _: QuantifierVariableLength => true
+        case _ => false
+      }
+    } else {
+      false
+    }
+  }
+
+  private def parseFactor(): RegexAST = {
+    var base = parseBase()
+    while (!eof() && (peek().exists(ch => ch == '*' || ch == '+' || ch == '?')
+        || isValidQuantifierAhead())) {
+
+      if (peek().contains('{')) {
+        consumeExpected('{')
+        base = RegexRepetition(base, parseQuantifierOrLiteralBrace().asInstanceOf[RegexQuantifier])
+      } else {
+        base = RegexRepetition(base, SimpleQuantifier(consume()))
+      }
```

Comment:
Nit: Suggest factoring out the common code:
```suggestion
      val quantifier = if (peek().contains('{')) {
        consumeExpected('{')
        parseQuantifierOrLiteralBrace().asInstanceOf[RegexQuantifier]\
      } else {
        SimpleQuantifier(consume())
      }
      base = RegexRepetition(base, quantifier)
```